test/optimizer/pushdown/filter_cannot_pushdown.test

```diff
@@ -0,0 +1,38 @@
+# name: test/optimizer/pushdown/filter_cannot_pushdown.test
+# description: Test Filter Can Not Push Down
+# group: [pushdown]
+
+require skip_reload
+
+# test some expressions which have side effects can not push down
+
+statement ok
+PRAGMA explain_output = OPTIMIZED_ONLY;
+
+query II
+explain select rnd from (select random()) as t(rnd) where rnd < 0.5;
+----
+logical_opt	<REGEX>:.*rnd < 0.5.*
+
+query II
+explain select * from (select rnd from (select random()) as t(rnd) where rnd < 0.5);
+----
+logical_opt	<REGEX>:.*rnd < 0.5.*
+
+statement ok
+CREATE TABLE t(a integer, b integer, c integer);
+
+query II
+explain select rnd, a from (select random() as rnd, a from t) where rnd < 0.3 and a > 1;
+----
+logical_opt	<REGEX>:.*rnd < 0.3.*
+
+query II
+explain select * from (select rnd, a from (select random() as rnd, a from t) where rnd < 0.3 and a > 1)
+----
+logical_opt	<REGEX>:.*rnd < 0.3.*
+
+query II
+explain select rnd, a from (select random() as rnd, a from t) where rnd < 0.3 and a > 1;
```

Comment:
might also add a test case like 

`explain select rnd, a from (select random(), 2 as 'a') as t(rnd, a) where rnd < 1 and a > 0;`

I wrote the following case as well 
```
create table t1 as select range as a, random() as b from range(10);
explain select a, b from (select random(), a, b from t1) as t(rnd, a, b) where rnd < 1 and a > 0;
```

And it seems to have a different plan than master duckdb. 
Master 
```
┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│           FILTER          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│      (random() < 1.0)     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           EC: 0           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         SEQ_SCAN          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             t1            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             a             │
│             b             │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│ Filters: a>0 AND a IS NOT │
│            NULL           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           EC: 0           │
└───────────────────────────┘
D
```

your branch 
```
┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             #1            │
│             #2            │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           FILTER          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        (rnd < 1.0)        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           EC: 10          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          random()         │
│             a             │
│             b             │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         SEQ_SCAN          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             t1            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             a             │
│             b             │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│ Filters: a>0 AND a IS NOT │
│            NULL           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           EC: 0           │
└───────────────────────────┘
```

Why is there now an extra projection for the random?

Reply:
I think this case is a little bit special. You can see my plan, you will find that if the filter not be pushed down below the projection, it will reference the column `rnd` and the top projection needs to eliminate the column `rnd`.


But you can see the master plan, if the filter can be pushed down below the projection. The origin logical plan will be `SQL_SCAN -> Filter(random()) -> PROJECTION(a, b)`. And the optimizer will analyze the plan, it will find the column `rnd` are not been used. So in the `SEQ_SCAN`, we don't need to extract the column `rnd`. We only needs the column `a` and column `b`. So the projection in the top is same, it also needs the column `a` and column `b`. So the projection can be removed. That's why the final plan in the master branch has no projection.

Reply:
@Tmonster Thanks for your comments. So, do you think the new plan is acceptable? And I'll add these test cases later.

Reply:
new plan is good